File: api/src/main/java/me/kvdpxne/dtm/Identifiable.java
package me.kvdpxne.dtm;

import java.io.Serializable;
import org.jetbrains.annotations.NotNull;

/**
 * Represents an entity that can be uniquely identified by an identifier.
 * <p>
 * The {@code Identifiable} interface provides a method to retrieve this
 * identifier, which must implement the {@link Serializable} interface.
 *
 * @param <T> the type of the identifier, which must extend
 *            {@link Serializable}.
 * @since 0.1.0
 */
public interface Identifiable<T extends Serializable> {

  /**
   * Retrieves the unique identifier for this entity.
   *
   * @return a non-null value of type {@code T} representing the identifier.
   * @since 0.1.0
   */
  @NotNull
  T getIdentifier();
}

File: src/main/kotlin/me/kvdpxne/dtm/shared/Identifiable.kt
package me.kvdpxne.dtm.shared

import java.io.Serializable

/**
 * Interface representing objects that can be uniquely identified by a
 * serializable value of type `T`.
 *
 * This interface allows for flexibility in the identifier type as long as
 * it implements `Serializable`. Common choices for T could be `String`, `Long`,
 * or custom serializable identifier classes.
 *
 * @param T The type of the identifier, which must be serializable.
 *
 * @since 0.1.0
 */
interface Identifiable<T : Serializable> {

  /**
   * The unique identifier of the object.
   *
   * @since 0.1.0
   */
  val identifier: T
}